src/index.ts

```diff
@@ -70,11 +70,22 @@ export function merge <T extends object, U extends object> (obj: T, src: U): T &
   for (const key in src) {
     const clone = isMergeableObject(src[key]) ? cloneDeep(src[key]) : undefined
     let x: (T & U)[Extract<keyof U, string>]
-    if (clone && isMergeableObject((x = res[key]))) {
+    // `has` check needed to avoid using inherited properties. This is probably
+    // closer to the behaviour one expects and also avoids prototype pollution.
+    if (clone && has(obj, key) && isMergeableObject((x = res[key]))) {
```

Comment:
> It'll merge the object in, but you're right that this is a bug because a clone of src[key] should be used. The correct location to do this check is (was) below to avoid merging on a property that obj doesn't have (like obj.__proto__ or obj.constructor)

I do not see how this addresses the concern.

If I have an object `o = { a: 5 }`, and another object `b = { b: 8 }`, then `merge(o, b)` should work to result in an updated `o` object that contains `{ a: 5, b: 8 }`.

But this code will prevent that from happening because `has(o, 'b')` will return `false` and therefore the key `'b'` won't be merged into `o`.

Could you also update the `merge` tests to check for this scenario so that they'll fail if this happens?

Reply:
> I do not see how this addresses the concern.

Before: this branch would still be skipped because there was a call to `isMergeableObject((x = res[key]))` which would have been `false` for `res['b']`.

Now: this branch would be skipped, as you mention, because of `has`, before even calling `isMergeableObject`.

Why it doesn't matter: then `Object.defineProperty` below is used.

 > Could you also update the merge tests to check for this scenario so that they'll catch this issue and fail?

Just did. Notice that `target` starts empty and ends up `{ __proto__: { ... } }`